### ISSUE TITLE
[BUG] Regarding the effectiveness of MovieLens retrieval.

### ISSUE THREAD

**Issue by ooooooover** [NVIDIA/recsys-examples#81]
**Describe the bug**
I found that there's an issue with the retrieval results on MovieLens. The hitrate@10 and NDCG@10 are as high as 0.98.

**Steps/Code to reproduce bug**
run PYTHONPATH=${PYTHONPATH}:$(realpath ../) torchrun --nproc_per_node 4 --master_addr localhost --master_port 6000  pretrain_gr_retrieval.py --gin-config-file movielen_retrieval.gin
**Expected behavior**

**Environment details (please complete the following information):**
 - Environment location: Docker
 - Method of recsys-examples install:Docker,

 

**Additional context**
I suspect that the issue arises because the embeddings generated during forward propagation are of shape [B*T, D] instead of [B, T, D].

**Comment by ooooooover** [comment]
<img width="613" alt="Image" src="https://github.com/user-attachments/assets/5815feb9-2fd5-46b8-b63f-c62b7f9af4cb" />

**Comment by ooooooover** [comment]
There is a logical error in the calculation of SampledSoftmaxLoss.
<img width="613" alt="Image" src="https://github.com/user-attachments/assets/c89f2164-5d98-49aa-bd1a-73141d40f13f" />

**Comment by JacoCheung** [comment]
Hi @ooooooover , thanks for your feedback. The docstring seems out of date. The supervisions embeddings and output_embeddings should be `[T, D]` where T is the total sequence length of current batch.

The rationale behind our loss compute is for **each** historical token, there is an `output embedding`, and a label  `supervision embedding`, and we sampled 128 `negative embeddings` from local batch to perform a softmax loss. So each token corresponds to a loss. The total loss should be identical regardless of the shape.

**Comment by ooooooover** [comment]
Why is the recall rate calculated on MovieLens as high as 0.98? Is there a reasonable explanation?

**Comment by ooooooover** [comment]
> Why is the recall rate calculated on MovieLens as high as 0.98? Is there a reasonable explanation?

<img width="912" alt="Image" src="https://github.com/user-attachments/assets/e2ea9f8c-3ac9-488a-954a-dfd554fd852f" />

**Comment by shijieliu** [comment]
hi @ooooooover we figured out the issue may be caused by incorrect data preprocessing(not sorting the samples) and we fixed it in #141 . Currently the log of the movielens retrieval tasks is 
```
[2025-09-02 09:29:10.366047] [train] [iter 99, tokens 1880465, elapsed_time 46876.60 ms, achieved FLOPS 1.57 TFLOPS]: loss 3.472153
325it [00:16, 19.54it/s]
[2025-09-02 09:29:50.668293] [eval] [eval 41600 users]:
    Metrics.NDCG@10:0.053441
    Metrics.NDCG@20:0.066035
    Metrics.HR@10:0.095489
[2025-09-02 09:29:57.238661] [train] [iter 199, tokens 1855279, elapsed_time 48241.81 ms, achieved FLOPS 1.45 TFLOPS]: loss 3.116923
178i
[2025-09-02 09:30:33.998447] [eval] [eval 41600 users]:
    Metrics.NDCG@10:0.071459
    Metrics.NDCG@20:0.088176
    Metrics.HR@10:0.127889
[2025-09-02 09:30:40.190725] [train] [iter 299, tokens 1845962, elapsed_time 42951.64 ms, achieved FLOPS 1.59 TFLOPS]: loss 2.797730
168it 
[2025-09-02 09:31:19.654878] [eval] [eval 41600 users]:
    Metrics.NDCG@10:0.081324
    Metrics.NDCG@20:0.100339
    Metrics.HR@10:0.146220
[2025-09-02 09:31:25.782350] [train] [iter 399, tokens 1795388, elapsed_time 45591.11 ms, achieved FLOPS 1.45 TFLOPS]: loss 2.737283
168it 
[2025-09-02 09:32:04.481608] [eval] [eval 41600 users]:
    Metrics.NDCG@10:0.088483
    Metrics.NDCG@20:0.108889
    Metrics.HR@10:0.159367
[2025-09-02 09:32:10.778375] [train] [iter 499, tokens 1869971, elapsed_time 44995.50 ms, achieved FLOPS 1.66 TFLOPS]: loss 2.708717
168it 
[2025-09-02 09:32:50.625092] [eval] [eval 41600 users]:
    Metrics.NDCG@10:0.093019
    Metrics.NDCG@20:0.114197
    Metrics.HR@10:0.168315
[2025-09-02 09:32:56.851680] [train] [iter 599, tokens 1813636, elapsed_time 46072.66 ms, achieved FLOPS 1.48 TFLOPS]: loss 2.566780
168it 
[2025-09-02 09:33:37.543839] [eval] [eval 41600 users]:
    Metrics.NDCG@10:0.096237
    Metrics.NDCG@20:0.118313
    Metrics.HR@10:0.174710
[2025-09-02 09:33:43.669449] [train] [iter 699, tokens 1811620, elapsed_time 46817.38 ms, achieved FLOPS 1.38 TFLOPS]: loss 2.482917
168it 
[2025-09-02 09:34:26.759178] [eval] [eval 41600 users]:
    Metrics.NDCG@10:0.101191
    Metrics.NDCG@20:0.123817
    Metrics.HR@10:0.183050
```
